pbj-core/pbj-grpc-helidon/docs/design.md

```diff
@@ -0,0 +1,455 @@
+        }
+        
+        return HelloReply.newBuilder().setMessage("Hello " + request.getName()).build();
+    }
+    
+    // ... other methods here
+}
+```
+
+Internally, the unary pipeline is implemented with a `Flow.Subscriber` that receives a single message from the client.
+It then uses the defined request convert to convert the message from `Bytes` to the appropriate type. The output of
+that call is then fed into the service implementation method, which returns the response. This response is then
+converted back to `Bytes` and sent back to the client.
+
+![Unary Pipeline](images/unary-call.png)
+
+## Server Streaming Pipeline
+
+The "server streaming" RPC method is one where a single message is sent from the client to the server, and the server
+responds with a stream of messages.
+
+Here is an example, where after receiving a message, the server responds with 10 messages. For this case, the method
+signature contains the request object and a `Flow.Subscriber`. PBJ handles converting the request object from `Bytes`
+into the appropriate type. The `Flow.Subscriber` is used to send messages back to the client, or terminate the
+connection due to errors, or to complete the connection gracefully.
+
+```java
+@Override
+public void sayHelloStreamReply(
+        @NonNull final HelloRequest request,
+        @NonNull final Flow.Subscriber<? super HelloReply> replies) {
+    for (int i = 0; i < 10; i++) {
+        replies.onNext(HelloReply.newBuilder()
+                .setMessage("Hello " + request.getName() + " " + i)
+                .build());
+    }
+
+    replies.onComplete();
+}
+```
+
+The server streaming pipeline is almost identical to the unary pipeline, except that the `replies` subscriber may be
+called multiple times, and is actually exposed to the application developer.
+
+![Server Streaming Pipeline](images/server-streaming.png)
+
+## Client Streaming Pipeline
+
+The "client streaming" RPC method is one where many messages are sent from the client to the server, but only a single
+message is sent back from the server to the client at the end.
+
+Here is an example. The client will send a stream of messages, each containing a name. The server will respond with a
+message containing the concatenation of all the names. Notice that the method must return a `Flow.Subscription` to
+the caller. This is the subscription to which the caller will push each message as it arrives from the client. Since
+we can only return a single object from the method, we need a callback into which to push the response message. We
+therefore see a `Flow.Subscription` supplied as an argument to the method, even though only a single message can
+be sent to the client. Sending more than one message will result in an error.
+
+```java
+@Override
+@NonNull
+public Flow.Subscriber<HelloRequest> sayHelloStreamRequest(@NonNull final Flow.Subscriber<HelloRequest> response) {
+    return new Flow.Subscriber<>() {
```

Comment:
nit: This might actually be easier to follow if the subscriber was not an anonymous inner class and response was passed via the constructor, it would certainly better highlight the simplicity of the stream request method in this case.

Reply:
I'm not sure I follow. Can you post a code example of what you have in mind?

Reply:
Just extract the new subscriber class to a static inner class so it's clearer how simple the subscriber method can be, and that the subscriber classes can (should) be reusable.
Something like this:
```java
@Override
@NonNull
public Flow.Subscriber<HelloRequest> sayHelloStreamRequest(@NonNull final Flow.Subscriber<HelloRequest> response) {
    return new JoinedEchoSubscriber(response);
}

/**
 * Simple subscriber that joins all of the streamed names together and
 * responds with a single "Hello" to all of those names joined together at
 * the end of the stream.
 */
private Static class JoinedEchoSubscriber extends Flow.Subscriber<HelloRequest> {
     private final List names = new ArrayList();
     private final Flow.Subscriber response;

     private JoinedEchoSubscriber(@NonNull final Flow.Subscriber<HelloRequest> response) {
         this.response = Objects.requireNonNull(response);
     }

     @Override
     public void onSubscribe(Flow.Subscription subscription) {
         subscription.request(Long.MAX_VALUE); // turn off flow control if you want
     }

     @Override
     public void onNext(HelloRequest item) {
         names.add(item.getName());
     }

     @Override
     public void onError(Throwable throwable) {
         response.onError(throwable);
     }

     @Override
     public void onComplete() {
         response.onNext(HelloReply.newBuilder()
                 .setMessage("Hello " + String.join(", ", names))
                 .build());
         response.onComplete();
     }
}
```